pipelined/bevy_render2/src/render_resource/pipeline_specializer.rs

```diff
@@ -20,15 +22,20 @@ impl<S: SpecializedPipeline> SpecializedPipelines<S> {
         cache: &mut RenderPipelineCache,
         specialize_pipeline: &S,
         key: S::Key,
+        vertex_layout: &VertexBufferLayout,
     ) -> CachedPipelineId {
         *self.cache.entry(key.clone()).or_insert_with(|| {
-            let descriptor = specialize_pipeline.specialize(key);
+            let descriptor = specialize_pipeline.specialize(key, vertex_layout);
```

Comment:
I haven't done a full review yet, but I wanted to call this out: I don't think this is "safe" to use as-is because the passed in vertex layout can affect the pipeline being constructed, but it isn't represented in the key/identity. The onus is on the caller to consistently pass in the correct (and identical) layout for a given key permutation.

Naively, we could just add the VertexBufferLayout to the key's hash, but I don't want to hash layouts every time. The best solution I can think of right now is to _not_ pass vertex buffer layouts into `specialize` at all. Instead, we could require "pre registering" specific vertex buffer layouts in the RenderPipelineCache, which would return a `VertexBufferLayoutKey` that serves as an identity for a specific layout. The RenderPipelineDescriptor could then be modified to accept an enum that either accepts an actual manually defined layout, or a VerterBufferLayoutKey. When the pipeline doesnt need to be re-configured based on arbitrary vertex inputs, we can continue using the approach we use today. But when we need to adapt to arbitrary layouts, we can then add VertexBufferLayoutKey to the specialization key, which could then be added to the descriptor as necessary. When the RenderPipelineCache actually compiles the descriptor, it would map the key to its internal collection of pre-registered layouts. Meshes, as part of being loaded, would pre-register their layouts (ideally in a way that uses the same key for identical layouts, to avoid ballooning memory).

Reply:
> Instead, we could require "pre registering" specific vertex buffer layouts in the RenderPipelineCache, which would return a `VertexBufferLayoutKey` that serves as an identity for a specific layout.

I attempted this approach, but it was a huge burden trying to pass the `VertexBufferLayout` to `RenderPipelineCache` from `Mesh`. It could just be me, but I couldn't identify a clean way to link these two together, since they currently do not know anything about one another. In short, I didn't want to tightly couple the pipeline cache with meshes. But perhaps that really is the right thing to do?

 The other thought I had while approaching this was maybe `Mesh` isn't the right owner for `VertexBufferLayout`? If the pipeline cache owns it (for instance) then the mesh needs a way to query the layout so it can produce the buffer. And that design has some other issues with coupling.

I'm definitely not happy with passing the the layout through `specialize()`, but it does seem like the cleanest and most simple solution at the moment. Especially since this feature is needed _right now_ and perfection can wait (IMHO).

Reply:
I agree that this works for the current use case (as long as you are _very_ careful about how you call `specialize()`). But this type of correctness issue isn't something I'm willing to brush aside. I certainly wouldn't call it "clean". It would be way too easy for users to pass in different layouts for the same key and get surprising results. Specializations need to be a pure function of their key and the SpecializedPipeline implementer state. Anything else is "broken".

I can't think of a better solution than pre-computed layout identities. They are cheap to include in keys and anything can "cache" them. Meshes probably shouldn't even care about them (directly). They can still just use the normal VertexBufferLayout api to produce a buffer. I think the `GpuMesh` is what should store the VertexBufferLayoutKey (if it is specific to the mesh). If it isn't specific to the mesh (ex: one of a few global values), we can store those in a resource or something. The value here is that we can cheaply store the key anywhere. I'm also open to any alternatives here. But I'm pretty sure I won't budge on "key correctness issues".

Reply:
> It would be way too easy for users to pass in different layouts for the same key and get surprising results. Specializations need to be a pure function of their key and the SpecializedPipeline implementer state. Anything else is "broken".

Yes, this makes a very strong case for finding a better solution. I don't know what it would look like just yet, but it is something I will be working on.

Reply:
@cart I have reverted this particular change. The vertex layouts are now cached on the `MeshPipeline` resource. It's a bit awkward to use now that the cache must be populated. You will see calls to `mesh_pipeline.cache_vertex_layout()` spread around in various places because of this.

I think the main concern with key correctness has been addressed. I can still make some improvements like incorporating the `VertexLayoutKey` idea, if you like.